### PR TITLE
am2rlauncher: init at 2.3.0

### PR DESCRIPTION
**Pull request by peeley**
## Description of changes

AM2RLauncher is a launcher for the Metroid 2 fan remake, AM2R. This program does not include the actual AM2R game itself; it only includes functionality for patching, updating, and modding the game.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/games/am2rlauncher/default.nix
Comment (SuperSamus) [inline pkgs/games/am2rlauncher/default.nix:19]: I don't usually see the function inputs formatted like this.
Use something like `alejandra` to format the derivation.

File: pkgs/games/am2rlauncher/default.nix
Comment (SuperSamus) [inline pkgs/games/am2rlauncher/default.nix:22]: I'd change the pname to `am2r-launcher` (same in `all-packages.nix`).
https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-naming

File: pkgs/games/am2rlauncher/default.nix
Comment (SuperSamus) [inline pkgs/games/am2rlauncher/default.nix:23]: Should be `unstable-2023-09-03` if you are using a non-tagged commit.

File: pkgs/games/am2rlauncher/default.nix
Comment (SuperSamus) [inline pkgs/games/am2rlauncher/default.nix:27]: That commit is out of tree, use this one instead.
5d8b7d9b3de68e6215c10b9fd223b7f1d5e40dea

(Also, change the sha256 accordingly. BTW, the current one isn't even correct)

File: pkgs/games/am2rlauncher/default.nix
Comment (SuperSamus) [inline pkgs/games/am2rlauncher/default.nix:98]: There is already a .desktop file in https://github.com/AM2R-Community-Developers/AM2RLauncher/blob/main/AM2RLauncher/distribution/linux/AM2RLauncher.desktop

The exec there is `AM2RLauncher` instead of `AM2RLauncher.Gtk`, though. The [AUR package](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=am2rlauncher) (linked in the README, I don't know if it's official) renames the executable to fit the .desktop file.

**Comment by SuperSamus** [comment]
While I didn't test the actual game patching and running, I noticed that when opening the launcher, you get two error pop-up windows. This is not good.

Also, the `Changelog` and `News` tabs don't work (they display: "TLS/SSL support not available; install glib-networking").
(I did try adding that dependency, and it didn't solve the issue, so I don't know how hard it is to fix)

**Review by SuperSamus — CHANGES_REQUESTED** [review]
While I didn't test the actual game patching and running, I noticed that when opening the launcher, you get two error pop-up windows. This is not good.

Also, the `Changelog` and `News` tabs don't work (they display: "TLS/SSL support not available; install glib-networking").

File: pkgs/games/am2rlauncher/default.nix
Comment (peeley) [reply pkgs/games/am2rlauncher/default.nix]: Unfortunately, it looks like NuGet doesn't play nicely with non-semver values of `version`. When I try `version = "unstable-2023-10-12`, I get the following build error:

```
error: builder for '/nix/store/qy41fia5d91nn3qfwa58y2hn005z3m83-am2r-launcher-unstable-2023-10-12.drv' failed with exit code 1;
       last 10 log lines:
       > unpacking sources
       > unpacking source archive /nix/store/hdgissird7hpfn8860dqkvazni29ykvm-source
       > source root is source
       > patching sources
       > applying patch /nix/store/66i1s7gz5zw0mpvndid6xr1l3bk8738l-steam-run-binary.patch
       > patching file AM2RLauncher/AM2RLauncherLib/Profile.cs
       > configuring
       > Executing dotnetConfigureHook
       >   Determining projects to restore...
       > /nix/store/dn7x9k9gy7xrnl92phh66i8j3yh0vj5s-dotnet-sdk-6.0.408/sdk/6.0.408/NuGet.targets(132,5): error : 'unstable-2023-10-12' is not a valid version string. (Parameter 'value') [/build/source/AM2RLauncher/AM2RLauncher.Gtk/AM2RLauncher.Gtk.csproj]
       For full logs, run 'nix log /nix/store/qy41fia5d91nn3qfwa58y2hn005z3m83-am2r-launcher-unstable-2023-10-12.drv'.
```

I can contact the maintainers and see if they are willing to do a minor version bump for this package, however.

**Comment by peeley** [comment]
@SuperSamus Apologies for such a long delay! I believe I've addressed most of the feedback, other than the `version` attribute change.

Concerning the error messages on startup, that is definitely concerning;  I've not experienced the same. Do you mind sharing what version of NixOS/nixpkgs you're on? Here's mine:

```
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.0.10, NixOS, 22.11 (Raccoon), 22.11.20221130.4d2b37a`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.11.0`
 - channels(root): `"nixos-22.05pre376605.f419dc5763c"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

$ nix registry list
user   flake:nixpkgs github:NixOS/nixpkgs/78058d810644f5ed276804ce7ea9e82d92bee293
```

**Comment by SuperSamus** [comment]
> Concerning the error messages on startup, that is definitely concerning

I still get it.
```
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.5.6-cachyos, NixOS, 23.11 (Tapir), 23.11.20231008.2646b29`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.17.0`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

I notice you are on NixOS 22.11, even though the latest stable is on 23.05.

Anyway, the actual error popups are (both are the same):

```
Unhandled Exception has occured in your Application.
*****Stack Trace*****

System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
---> System.UriFormatException: Invalid URI: The URI is empty.
at System.Uri.CreateThis(String uri, Boolean dontEscape, UriKind uriKind, UriCreationOptions& creationOptions)
at System.Uri..ctor(String uriString)
at Eto.GtkSharp.Forms.Controls.WebViewHandler.get_Url()
at
Eto.GtkSharp.Forms.Controls.WebViewHandler.WebViewHandler_LoadChanged(Object o, SignalArgs args)
--- End of inner exception stack trace ---
at System.RuntimeMethodHandle. InvokeMethod(Object target, Span`1& arguments, Signature sig, Boolean constructor, Boolean wrapExceptions) at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, Binding Flags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
at System.Delegate.DynamicInvokeImpl(Object[] args)
at System.Delegate.DynamicInvoke(Object[] args)
at GLib.Signal.ClosureInvoked CB(Object o, ClosureInvokedArgs args)
at GLib.SignalClosure.MarshalCallback(IntPtr raw_closure, IntPtr return_val, UInt32 n_param_vals, IntPtr param_values, IntPtr invocation_hint, IntPtr marshal_data)
```


Also, another thing I should have probably mentioned: in order to launch the executable, `steam-run` is used.
However, `steam-run` is a very heavy dependency (and it's also unfree). Because AM2RLauncher only launches AM2R (which has a predictable set of dependencies), it'd probably be better to use a `buildFHSEnv` (or some variant of it).

**Comment by peeley** [comment]
> I notice you are on NixOS 22.11, even though the latest stable is on 23.05.

Huh, even after upgrading to the following, I'm still having a hard time reproducing: 
```bash
$ nix-shell -p nix-info --run "nix-info -m"              
 - system: `"x86_64-linux"`
 - host os: `Linux 6.5.7, NixOS, 23.11 (Tapir), 23.11.20231011.01441e1`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.17.0`
 - channels(root): `"nixos-22.05pre376605.f419dc5763c"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Do you mind sharing how you're running the package? Also, could you share the contents of `~/.local/share/AM2RLauncher/Logs/AM2RLauncher.log`?  I've added `glib-networking` just as an experiment until I can reproduce or find the root cause; let me know if that makes any difference.

> Because AM2RLauncher only launches AM2R (which has a predictable set of dependencies), it'd probably be better to use a buildFHSEnv (or some variant of it).

The AM2R documentation uses `steam-run` in the directions for running on NixOS ([link](https://github.com/AM2R-Community-Developers/AM2R-Autopatcher-Linux#nix-including-nixos-1)). I can try to get `buildFHSEnv` working, but I predict that `steam-run` will be the more painless route.

**Comment by SuperSamus** [comment]
Taken on the current version here (and rebased to nixpkgs' 01441e14af5e). Not much to see...
[AM2RLauncher.log](https://github.com/NixOS/nixpkgs/files/12899821/AM2RLauncher.log)

**Comment by SuperSamus** [comment]
Back to the FHS environment, I made it myself. Cherry-pick it from https://github.com/SuperSamus/nixpkgs/tree/am2rlauncher-init.

**Comment by peeley** [comment]
Cherry-picked and pushed up the changes. I'm unfortunately still having a hell of a time reproducing the `Invalid URI` exception, even after running on a fresh install of NixOS on a separate machine.

I have a [branch](https://github.com/peeley/AM2RLauncher/tree/nixify) of AM2RLauncher that contains a `flake.nix` for running and testing local changes to the launcher. I'll try to introduce breaking changes to find the root cause of the exception.

**Comment by SuperSamus** [comment]
> In your NixOS configuration.nix (or corresponding flake), is `services.gnome.glib-networking.enable = true`?

It wasn't (it isn't enabled on KDE Plasma, the DE I'm using), and that was the culprit: the problems were fixed once I enabled it.
`glib-networking` can be removed from `runtimeDeps`, it doesn't do anything.

I think it'd be a good idea to open a PR to enable it by default on Plasma... (I'll do it)

**Comment by SuperSamus** [comment]
By the way, with the arrival of AM2R 1.6, the executable is switching to 64 bit, meaning that `am2r-run` needs to be tweaked. (After https://github.com/AM2R-Community-Developers/AM2RLauncher/pull/59 is merged)

```
  am2r-run = buildFHSEnv {
    name = "am2r-run";

    # 64 bit is for AM2R 1.6 and later.
    # 32 bit is for older versions and mods based on it.
    multiArch = true;
```

File: pkgs/games/am2rlauncher/default.nix
Comment (Miepee) [reply pkgs/games/am2rlauncher/default.nix]: FWIW, i'd personally rather have `am2rlauncher` than `am2r-launcher`. That's also the package name on the AUR.

File: pkgs/games/am2rlauncher/default.nix
Comment (Miepee) [reply pkgs/games/am2rlauncher/default.nix]: version bump is gonna prove a little difficult

Is `2.3.0-unstable-2023-09-03` acceptable?

Comment (peeley) [reply pkgs/games/am2rlauncher/default.nix]: `2.3.0-unstable-2023-09-03` does allow the package to build and run unlike `unstable-2023-09-03`, but I think it's unconventional for Nix packages. I'll defer to @SuperSamus on that one, since they have more packaging experience than myself.

File: pkgs/games/am2r-launcher/default.nix
Comment (Miepee) [inline pkgs/games/am2r-launcher/default.nix:22]: as someone not as familiar with nix, what does this wrapper do?

Comment (peeley) [reply pkgs/games/am2r-launcher/default.nix:22]: Essentially, it creates an isolated sandbox that follows the Linux Filesystem Hierarchy Standard. In Nix, all packages live in `/nix/store/` rather than in `/lib`, `/bin`, `/usr`, etc. When running a binary that expects that sort of environment and it isn't feasible to recompile or patch via `patchelf`, `buildFHSEnv` allows Nix to construct an environment that looks like a more "normal" Linux system and includes every library/package that the binary needs. This is most extensively used for games or proprietary software, where the code is closed-source and the binary isn't really modifiable.

In this example, we're making a sandbox that contains all the 32-bit packages that the game binary needs to run. The `am2r-run-binary.patch` file also patches the launcher so that the `runner` file is launched with the `am2r-run` script. This allows the binary to run in the FHS environment and find all of its dependencies.

Comment (SuperSamus) [reply pkgs/games/am2r-launcher/default.nix:22]: Just a reminder, this needs to be updated to support the 64 bit AM2R 1.6, https://github.com/NixOS/nixpkgs/pull/253203#issuecomment-1801405605.

File: pkgs/games/am2rlauncher/default.nix
Comment (SuperSamus) [reply pkgs/games/am2rlauncher/default.nix]: > `2.3.0-unstable-2023-09-03` does allow the package to build and run

Looking [here](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-naming), this seems to be allowed.

> Example: Given a project had its latest releases `2.2` in November 2021, and `3.0` in January 2022, a commit authored on March 15, 2022 for an upcoming bugfix release `2.2.1` would have `version = "2.2-unstable-2022-03-15"`.

File: pkgs/games/am2r-launcher/default.nix
Comment (Miepee) [reply pkgs/games/am2r-launcher/default.nix:22]: Ah thanks for the explanation. 
I did see the patch, and was wondering whether it makes sense to somehow get that into upstream. But considering that the nix-store is a very nix concept, it feels nicer having that as a patch.

File: pkgs/games/am2r-launcher/default.nix
Comment (SuperSamus) [inline pkgs/games/am2r-launcher/default.nix:3]: This can be removed.
(FYI, you can use language servers like `nil` to automatically detect unused inputs)

**Review by SuperSamus — APPROVED** [review]

File: pkgs/games/am2r-launcher/default.nix
Comment (Miepee) [inline pkgs/games/am2r-launcher/default.nix:46]: can this get changed to `am2rlauncher` please?

**Comment by SuperSandro2000** [comment]
> Also, the `Changelog` and `News` tabs don't work (they display: "TLS/SSL support not available; install glib-networking").

This means the wrapper is missing things.

**Comment by peeley** [comment]
> > Also, the `Changelog` and `News` tabs don't work (they display: "TLS/SSL support not available; install glib-networking").
> 
> This means the wrapper is missing things.

We were actually able to get to the root cause of this ([see here](https://github.com/NixOS/nixpkgs/pull/253203#issuecomment-1801400100)), and it appears to be unrelated. I believe there's a separate PR to fix the issue on KDE ([link](https://github.com/NixOS/nixpkgs/pull/266234)).

**Comment by SuperSamus** [comment]
Actually... I just found out that adding `--prefix GIO_EXTRA_MODULES : ${glib-networking.out}/lib/gio/modules` to `wrapProgram` fixes the issue.
The Plasma change is not necessary.
(See [here](https://github.com/SuperSamus/nixpkgs/blob/a914bdc551f12de2fec955b324946f96ca7f5600/pkgs/games/am2rlauncher/default.nix).)

File: pkgs/games/am2rlauncher/default.nix
Comment (SuperSandro2000) [inline pkgs/games/am2rlauncher/default.nix:83]:
```suggestion
      --prefix PATH : ${lib.makeBinPath [ am2r-run xdelta file busybox openjdk patchelf ]} \
```

Why do we need busybox here?

File: pkgs/games/am2rlauncher/default.nix
Comment (SuperSandro2000) [inline pkgs/games/am2rlauncher/default.nix:77]:
```suggestion
  dotnetFlags = [ ''-p:DefineConstants="NOAPPIMAGE;NOAUTOUPDATE;PATCHOPENSSL"'' ];
```

File: pkgs/games/am2rlauncher/default.nix
Comment (SuperSandro2000) [inline pkgs/games/am2rlauncher/default.nix:48]:
```suggestion
  version = "2.3.0-unstable-2023-11-08";

```
nit

File: pkgs/games/am2rlauncher/default.nix
Comment (SuperSandro2000) [inline pkgs/games/am2rlauncher/default.nix:39]:
```suggestion
    multiPkgs = pkgs: with pkgs; [
      stdenv.cc.cc.lib
      xorg.libX11
      xorg.libXext
      xorg.libXrandr
      xorg.libXxf86vm
      curl
      libGLU
      libglvnd
      openal
      zlib
    ];
```

**Comment by SuperSandro2000** [comment]
> By the way, with the arrival of AM2R 1.6, the executable is switching to 64 bit, meaning that `am2r-run` needs to be tweaked.

Do we really need 32 and 64 variants of the libs?

**Comment by peeley** [comment]
> > By the way, with the arrival of AM2R 1.6, the executable is switching to 64 bit, meaning that `am2r-run` needs to be tweaked.
> 
> Do we really need 32 and 64 variants of the libs?

The launcher allows the user to launch previous versions of the game (or arbitrary mods) that may be older versions and still depend on the 32-bit libs. For the user to be able to launch both the newest and previous versions of the game, both 32- and 64-bit variants of the libs are required.

File: pkgs/games/am2rlauncher/default.nix
Comment (peeley) [reply pkgs/games/am2rlauncher/default.nix]: Whoops sorry, that was left over from debugging the package. Removed.

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by peeley** [comment]
@SuperSandro2000 Sorry to bother you again - do you mind hitting the merge button if this is good to go? It appears that I don't have write access to the repository.

**Pull request merged**
